Minor update to the Error Handling docs

settermjd opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

This change:

- Completes some of the PHP code examples, so that it's clear that they refer to an entire file, not just a part of a file
- Provides an example for enabling the example `LoggingErrorListenerDelegatorFactory` used in the docs, for those unfamiliar with delegator factories (and those who are a little rusty with them 😄).

settermjd committed:
Minor update to the error handling docs

This change completes some of the PHP code examples, so that it's clear
that the example refers to an entire file, as well as provides an
example for enabling the example LoggingErrorListenerDelegatorFactory
used in the docs, for those unfamiliar with delegator factories (and
those who are a little rusty with them).

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

froschdesign commented on docs/book/v3/features/error-handling.md:
````diff
@@ -51,6 +51,8 @@ application:
 $app->pipe($errorMiddleware);
 ```
 
+## There is an easier way
````
Excerpt from the Writers Guide:

> For headlines the Chicago Manual of Style capitalization rules must be used:
> 
> 1. Capitalize the first and the last word.
> 1. Capitalize nouns, pronouns, adjectives, verbs, adverbs, and subordinate conjunctions.
> 1. Lowercase articles (a, an, the), coordinating conjunctions, and prepositions.
> 1. Lowercase the ‘to’ in an infinitive (I want to play guitar).
> 
> The online tool [capitalizemytitle.com](https://capitalizemytitle.com/style/Chicago/) can help to capitalize and case convert the headlines.
> 

```suggestion
## There Is an Easier Way
```

Some other suggestions for this section:

```suggestion
## Ready-Made Solution
```

```suggestion
## Use of the Ready-Made Handler
```

```suggestion
## Using the Prepared Handler
```

settermjd replied:
Thanks for the feedback (referring to all of it here). When did the Chicago Manual of Style become the default? Apologies if I suggested it and forgot. :-P

froschdesign commented on docs/book/v3/features/error-handling.md:
````diff
@@ -154,6 +156,9 @@ allowing the listener the ability to introspect the generated response as well.
 As an example, you could create a logging listener as follows:
 
 ```php
+<?php
+declare(strict_types=1);
+
````
Excerpt from the Writers Guide:

> To standardize the examples and reduce unnecessary repetition, the following rules must be observed:
> 
> - PHP's open and close tags are only allowed in code examples for view scripts
> - the declare construct must be omitted

settermjd replied:
Might it be worth adding a documented exception in this case? As always, I'm happy to go with the group consensus.

settermjd replied:
On reflection, you're right. Though, I'm keen to evolve the Writer's Guide.

froschdesign commented on docs/book/v3/features/error-handling.md:
````diff
@@ -195,6 +200,9 @@ You could then use a [delegator factory](container/delegator-factories.md) to
 create your logger listener and attach it to your error handler:
 
 ```php
+<?php
+declare(strict_types=1);
+
````
Excerpt from the Writers Guide:

> To standardize the examples and reduce unnecessary repetition, the following rules must be observed:
> 
> - PHP's open and close tags are only allowed in code examples for view scripts
> - the declare construct must be omitted

Ocramius commented on docs/book/v3/features/error-handling.md:
````diff
@@ -195,6 +200,9 @@ You could then use a [delegator factory](container/delegator-factories.md) to
 create your logger listener and attach it to your error handler:
 
 ```php
+<?php
````
IMO just noise, but I don't know our convention here

settermjd replied:
It could well be and I appreciate what you're saying. However, it's a convention that I've seen in most other online documentation and examples wrt PHP for some time now.

froschdesign replied:
@settermjd 
The problem here is that we don't have a definition to mark parts of code from files.

See in the documentation of Doctrine ORM:

This example is logical:

> ```php
> <?php
> // src/Bug.php
> 
> class Bug
> {
>     public function close()
>     {
>         $this->status = "CLOSE";
>     }
> }
> ```

https://www.doctrine-project.org/projects/doctrine-orm/en/current/tutorials/getting-started.html#updating-entities

But this one is not:

> ```php
> <?php
> $product = $entityManager->getRepository('Product')
>                          ->findOneBy(array('name' => $productName));
> ```

https://www.doctrine-project.org/projects/doctrine-orm/en/current/tutorials/getting-started.html#entity-repositories

---

Therefore, any suggestion is welcome. Please add it to the [documentation repository](https://github.com/laminas/documentation/issues).
Thanks in advance! 👍🏻

settermjd replied:
Will try and get this sorted in a day or two.

settermjd replied:
That's where a format such as AsciiDoc would come in handy, as blocks, especially code blocks, can include a filename and path, if required. Here's a blog post that covers it nicely: https://blog.mrhaki.com/2017/01/awesome-asciidoctor-using-filename.html. 

While not native to Markdown, there is a Mkdocs extension that might be of help: [python-markdown-extensions](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-a-title). If it is fully compatible with the Mkdocs version used to build the docs, then it might be a good solution. Thoughts, @froschdesign?

Ocramius commented on docs/book/v3/features/error-handling.md:
````diff
@@ -154,6 +156,9 @@ allowing the listener the ability to introspect the generated response as well.
 As an example, you could create a logging listener as follows:
 
 ```php
+<?php
````
IMO just noise, but I don't know our convention here

settermjd replied:
As above, mate.

Ocramius commented on docs/book/v3/features/error-handling.md:
````diff
@@ -213,6 +221,16 @@ class LoggingErrorListenerDelegatorFactory
+```php
+'delegators' => [
+    ErrorHandler::class => [
+        LoggingErrorListenerDelegatorFactory::class,
````
I'll be honest, the naming is really terrible (although pre-existing) 😬

settermjd replied:
True. I was focused on sticking as close to the existing documentation in this PR as possible. Happy to come back with a larger PR that reworks the naming used throughout this section of the docs.

settermjd replied:
Closing after accepting @froschdesign's update.

froschdesign commented on docs/book/v3/features/error-handling.md:
````diff
@@ -213,6 +221,16 @@ class LoggingErrorListenerDelegatorFactory
 }
 ```
 
+Then, enable the delegator factory in your application, such as by adding the following to the `getDependencies()` function in your app or module’s _ConfigProvider.php_ file.
````
Excerpt from the Writers Guide:

> The filename itself must be enclosed in backticks (`):
> 
> - for a single file: `filename.ext`
> - for a file in a directory: `directory/filename.ext`

```suggestion
Then, enable the delegator factory in your application, such as by adding the following to the `getDependencies()` function in your app or module’s `ConfigProvider.php` file.
```

froschdesign commented on docs/book/v3/features/error-handling.md:
````diff
@@ -213,6 +221,16 @@ class LoggingErrorListenerDelegatorFactory
 
+Then, enable the delegator factory in your application, such as by adding the following to the `getDependencies()` function in your app or module’s _ConfigProvider.php_ file.
+
+```php
+'delegators' => [
+    ErrorHandler::class => [
+        LoggingErrorListenerDelegatorFactory::class,
+    ]
+],
+```
````
I would extend the code example here to make the context clearer.

```suggestion
```php
public function getDependencies(): array
{
    return [
        'factories'  => [
            // …
        ],
        'delegators' => [
            ErrorHandler::class => [
                LoggingErrorListenerDelegatorFactory::class,
            ],
        ],
    ];
}
```
```

froschdesign reviewed (changes requested):
Nice improvement! 👍🏻 
I have added some comments.

settermjd committed:
Update docs/book/v3/features/error-handling.md

Co-authored-by: Frank Brückner <info@froschdesignstudio.de>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update docs/book/v3/features/error-handling.md

Co-authored-by: Frank Brückner <info@froschdesignstudio.de>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update docs/book/v3/features/error-handling.md

Co-authored-by: Frank Brückner <info@froschdesignstudio.de>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

alexmerlin commented:
Can we wrap this up? Only a couple of minor items left to adjust here...

settermjd commented:
> Can we wrap this up? Only a couple of minor items left to adjust here...

Will do.

settermjd committed:
Remove declare statements in the error handling docs

@froschdesign pointed out that these are not allowed by the Writer's
Guide. Removing them to comply with the guide.

gsteel reviewed (approved):
Thanks @settermjd :)
